### PR TITLE
<Community>: Add Initial Support for TiDB Vector Store

### PR DESCRIPTION
**Pull request by IANTHEREAL**
This pull request introduces initial support for the TiDB vector store. The current version is basic, laying the foundation for the vector store integration. While this implementation provides the essential features, we plan to expand and improve the TiDB vector store support with additional enhancements in future updates.

Upcoming Enhancements:
* Support for Vector Index Creation: To enhance the efficiency and performance of the vector store.
* Support for max marginal relevance search. 
* Customized Table Structure Support: Recognizing the need for flexibility, we plan for more tailored and efficient data store solutions.

Simple use case exmaple

```python
from typing import List, Tuple
from langchain.docstore.document import Document
from langchain_community.vectorstores import TiDBVectorStore
from langchain_openai import OpenAIEmbeddings

db = TiDBVectorStore.from_texts(
    embedding=embeddings,
    texts=['Andrew like eating oranges', 'Alexandra is from England', 'Ketanji Brown Jackson is a judge'],
    table_name="tidb_vector_langchain",
    connection_string=tidb_connection_url,
    distance_strategy="cosine",
)

query = "Can you tell me about Alexandra?"
docs_with_score: List[Tuple[Document, float]] = db.similarity_search_with_score(query)
for doc, score in docs_with_score:
    print("-" * 80)
    print("Score: ", score)
    print(doc.page_content)
    print("-" * 80)
```

### CODE REVIEW

**Comment by baskaryan** [comment]
@IANTHEREAL feel free to ping when the PR is ready for review

**Comment by IANTHEREAL** [comment]
@baskaryan The TiDB Vector feature is expected to release a beta version around the end of January. I will make the final adjustments then and invite you to review it 😊

File: libs/community/langchain_community/vectorstores/tidb_vector.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+DEFAULT_TiDB_VECTOR_STORE_NAME = "langchain_tidb_vector"
+
+
+class TiDBVector(VectorStore):
+    def __init__(
+        self,
+        connection_string: str,
+        embedding_function: Embeddings,
+        tidb_vectorstore: Optional[Any] = None,
+        vectorstore_name: str = DEFAULT_TiDB_VECTOR_STORE_NAME,
```
Comment (AV25242) [inline libs/community/langchain_community/vectorstores/tidb_vector.py:18]: Any reason why we dont want to call it table_name ? thats the convention when it comes to MySQL. 

By providing a default DEFAULT_TiDB_VECTOR_STORE_NAME are we saying that if a user doesn't provide a table name then a default table will be created ? If so what would be its structure ?

Comment (IANTHEREAL) [reply libs/community/langchain_community/vectorstores/tidb_vector.py]: addressed

**Comment by IANTHEREAL** [comment]
@hwchase17 @baskaryan 

Quick update on the TiDB Vector Store PR:

- Development Complete: The core functionality of TiDB Vector Store is developed and now in internal testing.
- Staging in February: A stable staging environment will be ready by February for thorough review.
- Launch & Documentation in March: Planning for a March release with comprehensive documentation, including Langchain + TiDB Vector use cases.

Feel free to reach out with any questions or concerns. Excited to see Langchain and TiDB Vector Store in action!

**Comment by IANTHEREAL** [comment]
Recently, there have been more and more code conflicts. If possible, please help review this PR.  @baskaryan @hwchase17   

And I've DM'd the access method for the tidb vector staging environment to @baskaryan, hope it's useful for this PR reviewing

**Review by baskaryan — APPROVED** [review]

**Comment by IANTHEREAL** [comment]
Thanks for reviewing @baskaryan

**Pull request merged**